staging/src/k8s.io/apimachinery/pkg/util/cache/expiring.go

```diff
@@ -0,0 +1,121 @@
+package cache
+
+import (
+	"sync"
+	"time"
+)
+
+type ExpiringDeleter interface {
+	Now() time.Time
+	DeleteAfter(delay time.Duration, del func()) (cancel func())
+}
+
+func NewExpiring() *Expiring {
+	return &Expiring{
+		deleter: realExpiringDeleter{},
+		m:       make(map[interface{}]entry),
+	}
+}
+
+// Expiring is a map whose entries expire after a per-entry timeout.
+type Expiring struct {
+	deleter ExpiringDeleter
+
+	// mu protects the below fields
+	mu sync.RWMutex
+	// m is the internal map that backs the cache.
+	m map[interface{}]entry
+	// generation is a generation number so that when setting multiple entries
+	// with the same key, the cleanup for earlier versions of the entry won't
+	// remove later ones.
+	generation uint64
+}
+
+type entry struct {
+	val        interface{}
+	expiry     time.Time
+	generation uint64
+	touch      func()
+}
+
+// Delete deletes an entry in the map.
+func (c *Expiring) Delete(key interface{}) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.del(key, 0)
+}
+
+// Get looks up an entry in the cache.
+func (c *Expiring) Get(key interface{}) (val interface{}, ok bool) {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	e, ok := c.m[key]
+	if !ok || !c.deleter.Now().Before(e.expiry) {
+		return nil, false
+	}
+	return e.val, true
+}
+
+// Set sets a key/value/expiry entry in the map, overwriting any previous entry
+// with the same key. The entry expires at the given expiry time, but it may be
+// removed earlier by other Set calls.
+func (c *Expiring) Set(key interface{}, val interface{}, ttl time.Duration) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	// Cancel cleanup.
+	c.del(key, 0)
+
+	expiry := c.deleter.Now().Add(ttl)
+
+	c.generation++
+	generation := c.generation
+
+	c.m[key] = entry{
+		val:        val,
+		expiry:     expiry,
+		generation: generation,
+		touch: c.deleter.DeleteAfter(ttl, func() {
+			c.mu.Lock()
+			defer c.mu.Unlock()
+			c.del(key, generation)
+		}),
+	}
+}
+
+// del deletes the entry for the given key. The generation argument is the
+// generation of the entry that should be deleted. If the generation has been
+// changed (e.g. if a set has occurred on an existing element but the old
+// cleanup still runs), this is a noop. If the generation argument is 0, the
+// entry's generation is ignored and the entry is deleted.
+func (c *Expiring) del(key interface{}, generation uint64) {
+	e, ok := c.m[key]
+	if !ok {
+		return
+	}
+	if generation != 0 && generation != e.generation {
+		return
+	}
+	e.touch()
+	delete(c.m, key)
+}
+
+// Len returns the number of items in the cache.
+func (c *Expiring) Len() int {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return len(c.m)
+}
+
+type realExpiringDeleter struct{}
+
+func (realExpiringDeleter) Now() time.Time {
+	return time.Now()
+}
+
+func (realExpiringDeleter) DeleteAfter(delay time.Duration, del func()) func() {
+	timer := time.AfterFunc(delay, del)
```

Comment:
I don't understand why this mechanism was chosen?

I think I was expecting periodic locks and mass deletions?

Possibly also combined with maintaining a sorted list of tokens by deletion time?

Running a timer per entry doesn't seem great?

Reply:
Ya, I touched on that in the description:

> The current implementation is intentionally as simple as possible. There are some possible performance improvements that add complexity:
>
> * GC of expired entrys can be combined so that a group of cleanups only need to acquire the exclusive lock once.
> * ExpiringDeleter can be implemented with a workqueue.DelayingInterface so that GC is performed on a single persistent goroutine. Right now a goroutine is spawned per cleanup.
> * The GC can be delayed for an arbitrary duration after the entry expires. For frequently accessed entries, this allows the entry to be reset rather than GC'd which reduces the number of exclusive locks that need to be acquired.

I was mostly trying to prove this ou[t to compare performance of the map access with the old implementation, then optimize GC. I suppose we should address all of these before switching implementations though. Would you agree?

Reply:
Actually I'm warming up to this mechanism a little bit, but:

* how high does the go timer mechanism scale? I know they've improved it, but I'm not sure if running 10,000's of high-churn timers will be a problem?
* a possible extension to make this safer for malicious tokens is make the deleter track a virtual time that trails Now() by a smaller amount when there's many tokens in the cache; I don't think this would be easy while using real timers.

Actually I have other ideas on handling malicious tokens but that's a problem for another day.

Reply:
Second is probably the biggest improvement one given that AfterFunc creates a goroutine per call which is a lot of garbage.

Reply:
> ExpiringDeleter can be implemented with a workqueue.DelayingInterface

Ah, OK, I missed that line. That would be a large improvement.

> I suppose we should address all of these before switching implementations though. Would you agree?

I think we need to be sure we aren't introducing new failure modes as bad as the ones we're fixing :)